existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,194 @@
+``` 
+
+Quite long right?
+This is because ghc will maintain the same order, as these variables first appear in function definition.
+However, using explicit `forall.` this can be simplified:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
+This is useful not only in such big examples, but also when your function requires any of its arguments to be specified on usage.
+
+#### Additional type variables
+Second, you now can define functions with more type variables than you use on the right side of declaration.
+This is used mostly when you want to add complex constraint, which require several types, without adding them into the definition itself:
+
+``` Haskell
+someMonadicFunction :: forall a m env. MonadReader env m => a -> m a
+```
+
+Here we don't have `env` used in the type definition, but despite that, we are still able to add `MonadReader` constraint.
+
+## Existential quantification
+Beside universally quantified variables, Haskell also supports existential types.
+In order to use them ... you still need to use `forall.` keyword ¯\\\_(ツ)_/¯.
+That might be a little confusing.
+The following construction is called existential type in Haskell - when you use a forall quantifier in the right-hand side of function or data type declaration.
+But what is it used for? Let's find out!
+
+### Usage
+
+We will start with functions:
+
+```haskell
+func :: forall a c. a -> (forall b. (Show b) => b) -> c
+```
+Few things to mention.
+Here `a` and `c` are universally quantified and `b` is existentially quantified variables.
+The same variable can't be quantified more than once.
+Also, you can see that constraints for `b` are added in the middle of the function type declaration.
+This example is pretty useless.
+The only thing that has changed here is that now you can't manually choose type for `b` - the compiler must deduce the correct variable you want to use by itself.
+
+When we come to data type declaration, it becomes more useful.
+
+#### Hiding type variables
+```Haskell
+data MyData a = forall b. MkMyData b a
+```
+The first thing you might have noticed is that we don't have a type variable for `b` on the left side of expression.
+However, we still can use it on the right side.
+This is because `b` is existentially quantified.
+
+```Haskell
+MkMyData @_ @Integer -- This won't work, you don't have second variable in declaration
+```
+Compiler will deduce the type of the variable itself, and you can't specify it.
+
+That is actually why this construction is called existential type - because you tell a compiler, that there exists type, that
+can be used in constructor, without specifying which one exactly. Also it is called so, because from type theory's point of view
```

Comment:
> because you tell a compiler, that there exists type, that
> can be used in constructor, without specifying which one exactly

I'm not sure this is very elucidating.

If had written this:

```hs
data MyData a b = MkMyData a b
```

I would also be telling the compiler that "there exists a type that can be used in the constructor, without specifying which one exactly".
```